handlers_test.go

```diff
@@ -91,3 +92,111 @@ func Test_GivenCorrectId_HandlerReturnsDeviceObject(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, expected, result)
 }
+
+func Test_GetAllDevicesHandler_GivenWrongInput_HandlerReturns400(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	urls := []string{"/devices?limit=a", "/devices?page=a",
+		"/devices?limit=1&page=a", "/devices?limit=-1",
+		"/devices?page=-1"}
+	for _, url := range urls {
+		resp, err := http.Get(mockServer.URL + url)
+
+		assert.NoError(t, err)
+		assert.Equal(t, http.StatusBadRequest, resp.StatusCode)
+	}
+}
+
+func Test_GetAllDevicesHandler_NoParams_HandlerDefaultsLimitAndPage(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+}
+
+func Test_GetAllDevicesHandler_GivenDaoError_HandlerReturns500(t *testing.T) {
+	r := newRouter(NewService(&mockDao{returnErr: ErrDao("")}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+}
+
+func Test_GetAllDevicesHandler_GivenLimitZero_HandlerReturnsAllDevices(t *testing.T) {
+	m := &mockDao{data: make(map[int]Device)}
+	m.data[0] = Device{Name: "test name"}
+	r := newRouter(NewService(m))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices?limit=0")
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+
+	expected := []Device{{Name: "test name"}}
+
+	var result []Device
+	err = json.NewDecoder(resp.Body).Decode(&result)
+
+	fmt.Println(result)
```

Comment:
Leaving `printlines` in the code is generally a bad practice. In tests you can use `t.Log` - this func prints to the console but keeps it clean unless a test has failed.

Reply:
I forgot about this one, leaving it there wasn't intentional, will use log instead in the future though